src/options/options.h

```diff
@@ -415,6 +419,13 @@ class CVC4_PUBLIC Options {
   void flushOut();
 
  private:
+  /** Pointer to the options listener, if one exists */
+  OptionsListener* d_olisten;
```

Comment:
I'm somewhat confused here. You added a generic `OptionsListener` class.
- Is this supposed to replace all the old listeners that we have right now? (`d_beforeSearchListeners` et al)
- As this is generic, I'd suggest using (the equivalent of) `ListenerCollection` here to allow for multiple listeners. If we know there will only ever by a single one, why not use a `OptionsManager*` directly and avoid this layer of (conceptual) indirection?

Reply:
Yes, this replaces all listeners. 

Concerning ListenerCollection, I am also planning to delete this class entirely. I believe that this is overdesigned: if the user of this class wants to register multiple listeners, then this is equivalent to implementing multiple things in their (single) listener.  The only reason for having a listener collection is if multiple independent entities want to listen to the same options object, which is not the case. 

The level of indirection here is to avoid dependencies.  The code for "implementing" the options (setting output, resources) should not live in the options code itself.  This separates the two and allows all code to live in `src/smt/`. Indeed its true that we may only have a single listener, but that was the purpose of this.

Reply:
I see. Note that the `ListenerCollection` is also used in the `ResourceManager`. Similarly, the `SmtEngine` seems to be the only entity registering a listener there right now. We planned to allow registering listeners (for resource out, I think) via the API...
We may thus need (something like) the `ListenerCollection`after all. But I agree that the current implementation seems overdesigned...

Reply:
Yes, I managed to clean up resource manager in this regard as well. For reference, here's my branch: 

https://github.com/CVC4/CVC4/compare/master...ajreynol:refactorOptListen

I would prefer designing the new listener stuff from scratch if it is needed for the API.  I would still argue that this doesn't have to impact the listener in options.  That is, I would prefer registering multiple user-provided callbacks in the derived OptionsListener to separate the complexity of these things.  This way options just has to make one single callback.

Reply:
Okay.